presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java

```diff
@@ -1507,6 +1509,155 @@ private static void validateSphericalType(String function, OGCGeometry geometry,
+        {
+            previousPhi = toRadians(endPoint.getY());
+            previousSin = Math.sin(previousPhi);
+            previousCos = Math.cos(previousPhi);
+            previousTan = Math.tan(previousPhi / 2);
+            previousLongitude = toRadians(endPoint.getX());
+            firstPoint = true;
+        }
+
+        private void add(Point point)
+        {
+            double phi = toRadians(point.getY());
+            double tan = Math.tan(phi / 2);
+            double longitude = toRadians(point.getX());
+
+            // Only process points which are different from the previous one,
+            // Otherwise calculating the bearing is not deterministic
+            if (longitude != previousLongitude || phi != previousPhi) {
+                double deltaLongitude = longitude - previousLongitude;
+                sphericalExcess += 2 * Math.atan2(Math.tan(deltaLongitude / 2) * (previousTan + tan), 1 + previousTan * tan);
+
+                double cos = Math.cos(phi);
+                double sin = Math.sin(phi);
+                double sinOfDeltaLongitude = Math.sin(deltaLongitude);
+                double cosOfDeltaLongitude = Math.cos(deltaLongitude);
+
+                // Initial bearing from previous to current
+                double y = sinOfDeltaLongitude * cos;
+                double x = previousCos * sin - previousSin * cos * cosOfDeltaLongitude;
+                double initialBearing = (Math.atan2(y, x) + TWO_PI) % TWO_PI;
+
+                // Final bearing from previous to current = opposite of bearing from current to previous
+                double finalY = -sinOfDeltaLongitude * previousCos;
+                double finalX = previousSin * cos - previousCos * sin * cosOfDeltaLongitude;
+                double finalBearing = (Math.atan2(finalY, finalX) + PI) % TWO_PI;
+
+                // When processing our first point we don't yet have a previousFinalBearing
+                if (firstPoint) {
+                    // So keep our initial bearing around, and we'll use it at the end
+                    // with the last final bearing
+                    firstInitialBearing = initialBearing;
+                    firstPoint = false;
+                }
+                else {
+                    courseDelta += (initialBearing - previousFinalBearing + THREE_PI) % TWO_PI - PI;
+                }
+
+                courseDelta += (finalBearing - initialBearing + THREE_PI) % TWO_PI - PI;
+
+                previousFinalBearing = finalBearing;
+                previousCos = cos;
+                previousSin = sin;
+                previousPhi = phi;
+                previousTan = tan;
+                previousLongitude = longitude;
+            }
+        }
+
+        public double getSphericalExcess()
+        {
+            // Now that we have our last final bearing, we can calculate the remaining course delta
+            courseDelta += (firstInitialBearing - previousFinalBearing + THREE_PI) % TWO_PI - PI;
```

Comment:
This logic mutates internal state and makes it illegal to invoke `getSphericalExcess` multiple times. Perhaps, add a boolean to enforce that getSphericalExcess is called only once and add is never called after getSphericalExcess? It might be clearer to rename `getSphericalExcess` to `computeSphericalExcess`.